File: macos/DotsHarness/Tests/HarnessPluginKitTests/ProjectRulesTests.swift
import XCTest
@testable import DotsHarnessCore

final class ProjectRulesTests: XCTestCase {
    private var workspace: URL!

    override func setUpWithError() throws {
        workspace = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("ProjectRulesTests-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: workspace, withIntermediateDirectories: true)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: workspace)
    }

    private func write(_ relative: String, _ contents: String) throws {
        let url = workspace.appendingPathComponent(relative)
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try contents.write(to: url, atomically: true, encoding: .utf8)
    }

    func testReadsKnownRulesFilesAndLabelsThem() throws {
        try write("AGENTS.md", "use tabs")
        try write(".cursor/rules/style.mdc", "no semicolons")

        let text = ProjectRules.text(workspace: workspace)

        XCTAssertTrue(text.contains("# AGENTS.md\nuse tabs"))
        XCTAssertTrue(text.contains("# .cursor/rules/style.mdc\nno semicolons"))
        // AGENTS.md is the general file, so it leads.
        XCTAssertTrue(text.hasPrefix("# AGENTS.md"))
    }

    func testNewerRulesFileIsMirroredOntoTheOtherAndEmittedOnce() throws {
        try write("CLAUDE.md", "stale")
        try FileManager.default.setAttributes(
            [.modificationDate: Date().addingTimeInterval(-300)],
            ofItemAtPath: workspace.appendingPathComponent("CLAUDE.md").path
        )
        try write("AGENTS.md", "use tabs")

        let text = ProjectRules.text(workspace: workspace)

        let mirrored = try String(contentsOf: workspace.appendingPathComponent("CLAUDE.md"), encoding: .utf8)
        XCTAssertEqual(mirrored, "use tabs")
        XCTAssertEqual(text, "# AGENTS.md\nuse tabs")
    }

    func testMissingCounterpartIsCreated() throws {
        try write("CLAUDE.md", "run make test")
        ProjectRules.sync(workspace: workspace)
        let created = try String(contentsOf: workspace.appendingPathComponent("AGENTS.md"), encoding: .utf8)
        XCTAssertEqual(created, "run make test")
    }

    func testEmptyWorkspaceIsSeededWithBothRulesFiles() throws {
        XCTAssertTrue(ProjectRules.seedIfMissing(workspace: workspace))

        for name in ["AGENTS.md", "CLAUDE.md"] {
            XCTAssertEqual(
                try String(contentsOf: workspace.appendingPathComponent(name), encoding: .utf8),
                ProjectRules.seed
            )
        }
        XCTAssertEqual(ProjectRules.text(workspace: workspace), "# AGENTS.md\n" + ProjectRules.seed)
    }

    func testSeedLeavesAProjectThatAlreadyHasEitherFileAlone() throws {
        try write("CLAUDE.md", "run make test")
        XCTAssertFalse(ProjectRules.seedIfMissing(workspace: workspace))
        XCTAssertFalse(FileManager.default.fileExists(atPath: workspace.appendingPathComponent("AGENTS.md").path))
        XCTAssertEqual(
            try String(contentsOf: workspace.appendingPathComponent("CLAUDE.md"), encoding: .utf8),
            "run make test"
        )
    }

    func testTextNeverSeeds() throws {
        XCTAssertEqual(ProjectRules.text(workspace: workspace), "")
        XCTAssertFalse(FileManager.default.fileExists(atPath: workspace.appendingPathComponent("AGENTS.md").path))
    }

    func testNoWorkspaceAndBlankRulesFileYieldNothing() throws {
        XCTAssertEqual(ProjectRules.text(workspace: nil), "")
        try write("AGENTS.md", "   \n  ")
        XCTAssertEqual(ProjectRules.text(workspace: workspace), "")
    }

    func testOversizedFileIsTruncated() throws {
        try write("AGENTS.md", String(repeating: "x", count: ProjectRules.maxBytesPerFile * 2))
        let text = ProjectRules.text(workspace: workspace)
        XCTAssertTrue(text.hasSuffix("… truncated"))
        XCTAssertLessThan(text.utf8.count, ProjectRules.maxBytesPerFile + 200)
    }

    func testSymlinkEscapingTheWorkspaceIsIgnored() throws {
        let outside = workspace.deletingLastPathComponent()
            .appendingPathComponent("outside-\(UUID().uuidString).md")
        try "secret rules".write(to: outside, atomically: true, encoding: .utf8)
        defer { try? FileManager.default.removeItem(at: outside) }
        try FileManager.default.createSymbolicLink(
            at: workspace.appendingPathComponent("AGENTS.md"),
            withDestinationURL: outside
        )

        XCTAssertEqual(ProjectRules.text(workspace: workspace), "")
    }
}

File: shared/ProjectRulesTests.cs
using DotsHarnessCore;
using Xunit;

public sealed class ProjectRulesTests : IDisposable
{
    private readonly string _workspace = Path.Combine(
        Path.GetTempPath(),
        $"ProjectRulesTests-{Guid.NewGuid():N}");

    public ProjectRulesTests() => Directory.CreateDirectory(_workspace);

    public void Dispose()
    {
        try { Directory.Delete(_workspace, true); } catch (IOException) { }
    }

    private void Write(string relative, string contents)
    {
        var path = Path.Combine(_workspace, relative.Replace('/', Path.DirectorySeparatorChar));
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        File.WriteAllText(path, contents);
    }

    [Fact]
    public void ReadsKnownRulesFilesAndLabelsThem()
    {
        Write("AGENTS.md", "use tabs");
        Write(".cursor/rules/style.mdc", "no semicolons");

        var text = ProjectRules.Text(_workspace);

        Assert.Contains("# AGENTS.md\nuse tabs", text);
        Assert.Contains("# .cursor/rules/style.mdc\nno semicolons", text);
        // AGENTS.md is the general file, so it leads.
        Assert.StartsWith("# AGENTS.md", text);
    }

    [Fact]
    public void NewerRulesFileIsMirroredOntoTheOtherAndEmittedOnce()
    {
        Write("AGENTS.md", "use tabs");
        Write("CLAUDE.md", "stale");
        File.SetLastWriteTimeUtc(Path.Combine(_workspace, "CLAUDE.md"), DateTime.UtcNow.AddMinutes(-5));

        var text = ProjectRules.Text(_workspace);

        Assert.Equal("use tabs", File.ReadAllText(Path.Combine(_workspace, "CLAUDE.md")));
        Assert.Equal("# AGENTS.md\nuse tabs", text);
    }

    [Fact]
    public void MissingCounterpartIsCreated()
    {
        Write("CLAUDE.md", "run make test");
        ProjectRules.Sync(_workspace);
        Assert.Equal("run make test", File.ReadAllText(Path.Combine(_workspace, "AGENTS.md")));
    }

    [Fact]
    public void EmptyWorkspaceIsSeededWithBothRulesFiles()
    {
        Assert.True(ProjectRules.SeedIfMissing(_workspace));
        Assert.Equal(ProjectRules.Seed, File.ReadAllText(Path.Combine(_workspace, "AGENTS.md")));
        Assert.Equal(ProjectRules.Seed, File.ReadAllText(Path.Combine(_workspace, "CLAUDE.md")));
        Assert.Equal("# AGENTS.md\n" + ProjectRules.Seed, ProjectRules.Text(_workspace));
    }

    [Fact]
    public void SeedLeavesAProjectThatAlreadyHasEitherFileAlone()
    {
        Write("CLAUDE.md", "run make test");
        Assert.False(ProjectRules.SeedIfMissing(_workspace));
        Assert.False(File.Exists(Path.Combine(_workspace, "AGENTS.md")));
        Assert.Equal("run make test", File.ReadAllText(Path.Combine(_workspace, "CLAUDE.md")));
    }

    [Fact]
    public void TextNeverSeeds()
    {
        Assert.Equal("", ProjectRules.Text(_workspace));
        Assert.False(File.Exists(Path.Combine(_workspace, "AGENTS.md")));
    }

    [Fact]
    public void NoWorkspaceAndBlankRulesFileYieldNothing()
    {
        Assert.Equal("", ProjectRules.Text(null));
        Write("AGENTS.md", "   \n  ");
        Assert.Equal("", ProjectRules.Text(_workspace));
    }

    [Fact]
    public void OversizedFileIsTruncated()
    {
        Write("AGENTS.md", new string('x', ProjectRules.MaxBytesPerFile * 2));
        var text = ProjectRules.Text(_workspace);
        Assert.EndsWith("… truncated", text);
        Assert.True(text.Length < ProjectRules.MaxBytesPerFile + 200);
    }

    [Fact]
    public void SymlinkEscapingTheWorkspaceIsIgnored()
    {
        var outside = Path.Combine(Path.GetTempPath(), $"outside-{Guid.NewGuid():N}.md");
        File.WriteAllText(outside, "secret rules");
        try
        {
            File.CreateSymbolicLink(Path.Combine(_workspace, "AGENTS.md"), outside);
            Assert.Equal("", ProjectRules.Text(_workspace));
        }
        finally
        {
            File.Delete(outside);
        }
    }
}
